src/refresh/RefreshTimer.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ * The backoff time is between the minimum and maximum backoff time, based on the number of attempts.
+ * An exponential backoff strategy is used, with a jitter factor to prevent clients from retrying at the same time.
+ *
+ * The backoff time is calculated as follows:
+ *  - `basic backoff time` = `MinimumBackoffInMs` * 2 ^ `attempts`, and it is no larger than the `MaximumBackoffInMs`.
+ *  - based on jitter ratio, the jittered time is between [-1, 1) * `JitterRatio` * basic backoff time.
+ *  - the final backoff time is the basic backoff time plus the jittered time.
+ *
+ * Note: the backoff time usually is no larger than the refresh interval, which is specified by the user.
+ *  - If the interval is less than the minimum backoff, the interval is used.
+ *  - If the interval is between the minimum and maximum backoff, the interval is used as the maximum backoff.
+ *  - Because of the jitter, the maximum backoff time is actually `MaximumBackoffInMs` * (1 + `JitterRatio`).
+ */
+
+const MinimumBackoffInMs = 30 * 1000; // 30s
+const MaximumBackoffInMs = 10 * 60 * 1000; // 10min
+const MaxAttempts = 63;
+const JitterRatio = 0.25;
+
+export class RefreshTimer {
+    private _minBackoff: number;
+    private _maxBackoff: number;
+    private _attempts: number;
+    private _backoffEnd: number; // Timestamp
+    constructor(
+        private _interval: number

```

Comment:
What will happen if _interval is negative?

Reply:
Interval cannot be set to negative. Its min value is 1000ms.

Reply:
It seems like the python provider lacks the check for invalid refresh interval. I will talk to Matthew offline.

Reply:
But I agree that in terms of the RefreshTimer itself, we should check the interval passed into constructor and ensure it's positive.

Reply:
fixed in e0c9736